compiler/src/dmd/constfold.d

```diff
@@ -133,89 +133,121 @@ UnionExp Add(const ref Loc loc, Type type, Expression e1, Expression e2)
     static if (LOG)
     {
         printf("Add(e1 = %s, e2 = %s)\n", e1.toChars(), e2.toChars());
+        printf("type %s t1 %s t2 %s\n", type.toChars(), e1.type.toChars(), e2.type.toChars());
     }
-    if (type.isreal())
-    {
-        emplaceExp!(RealExp)(&ue, loc, e1.toReal() + e2.toReal(), type);
-    }
-    else if (type.isimaginary())
-    {
-        emplaceExp!(RealExp)(&ue, loc, e1.toImaginary() + e2.toImaginary(), type);
-    }
-    else if (type.iscomplex())
+
+    real_t e1re() { return e1.toReal(); }
+    real_t e1im() { return e1.toImaginary(); }
+    real_t e2re() { return e2.toReal(); }
+    real_t e2im() { return e2.toImaginary(); }
```

Comment:
I don't see the point for these functions.  `e.toReal` is much nicer to read, but if you insist on doing this you could just have 2 functions `re` and `im` that receive the expression as a parameter. Then you could simply do `e1.re`, `e2.im` etc.

Reply:
I tried several different ways, and this way made the use of them the easiest to read when used. The uses are all different and have to be carefully checked for being correct.